### PR TITLE
[asan][windows] Eliminate the static asan runtime on windows

### PR DESCRIPTION
**Pull request by barcharcraz**
This is one of the major changes we (Microsoft) have made in the version of asan we ship with Visual Studio. 

@amyw-msft wrote a blog post outlining this work at https://devblogs.microsoft.com/cppblog/msvc-address-sanitizer-one-dll-for-all-runtime-configurations/

> With Visual Studio 2022 version 17.7 Preview 3, we have refactored the MSVC Address Sanitizer (ASan) to depend on one runtime DLL regardless of the runtime configuration. This simplifies project onboarding and supports more scenarios, particularly for projects statically linked (/MT, /MTd) to the C Runtimes. However, static configurations have a new dependency on the ASan runtime DLL.

> Summary of the changes:

> ASan now works with /MT or /MTd built DLLs when the host EXE was not compiled with ASan. This includes Windows services, COM components, and plugins.
Configuring your project with ASan is now simpler, since your project doesn’t need to uniformly specify the same [runtime configuration](https://learn.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-170) (/MT, /MTd, /MD, /MDd).
ASan workflows and pipelines for /MT or /MTd built projects will need to ensure the ASan DLL (clang_rt.asan_dynamic-<arch>.dll) is available on PATH.
The names of the ASan .lib files needed by the linker have changed (the linker normally takes care of this if not manually specifying lib names via /INFERASANLIBS)
You cannot mix ASan-compiled binaries from previous versions of the MSVC Address Sanitizer (this is always true, but especially true in this case).

Here's the description of these changes from our internal PR

1. Build one DLL that includes everything debug mode needs (not included here, already contributed upstream).
* Remove #if _DEBUG checks everywhere.
* In some places, this needed to be replaced with a runtime check. In asan_win.cpp, IsDebugRuntimePresent was added where we are searching for allocations prior to ASAN initialization.
* In asan_win_runtime_functions.cpp and interception_win.cpp, we need to be aware of debug runtime DLLs even when not built with _DEBUG.
2. Redirect statically linked functions to the ASAN DLL for /MT
* New exports for each of the C allocation APIs so that the statically linked portion of the runtime can call them (see asan_malloc_win.cpp, search MALLOC_DLL_EXPORT). Since we want our stack trace information to be accurate and without noise, this means we need to capture stack frame info from the original call and tell it to our DLL export. For this, I have reused the __asan_win_new_delete_data used for op new/delete support from asan_win_new_delete_thunk_common.h and moved it into asan_win_thunk_common.h renamed as __asan_win_stack_data.
* For the C allocation APIs, a new file is included in the statically-linked /WHOLEARCHIVE lib - asan_malloc_win_thunk.cpp. These functions simply provide definitions for malloc/free/etc to be used instead of the UCRT's definitions for /MT and instead call the ASAN DLL export. /INFERASANLIBS ensures libucrt.lib will not take precedence via /WHOLEARCHIVE.
* For other APIs, the interception code was called, so a new export is provided: __sanitizer_override_function. __sanitizer_override_function_by_addr is also provided to support __except_handler4 on x86 (due to the security cookie being per-module).
3. Support weak symbols for /MD
* We have customers (CoreCLR) that rely on this behavior and would force /MT to get it.
* There was sanitizer_win_weak_interception.cpp before, which did some stuff for setting up the .WEAK section, but this only worked on /MT. Now stuff registered in the .WEAK section is passed to the ASAN DLL via new export __sanitizer_register_weak_function (impl in sanitizer_win_interception.cpp). Unlike linux, multiple weak symbol registrations are possible here. Current behavior is to give priority on module load order such that whoever loads last (so priority is given to the EXE) will have their weak symbol registered.
* Unfortunately, the registration can only occur during the user module startup, which is after ASAN DLL startup, so any weak symbols used by ASAN during initialization will not be picked up. This is most notable for __asan_default_options and friends (see asan_flags.cpp). A mechanism was made to add a callback for when a certain weak symbol was registered, so now we process __asan_default_options during module startup instead of ASAN startup. This is a change in behavior, but there's no real way around this due to how DLLs are.
4. Build reorganization
* I noticed that our current build configuration is very MSVC-specific and so did a bit of reworking. Removed a lot of create_multiple_windows_obj_lib use since it's no longer needed and it changed how we needed to refer to each object_lib by adding runtime configuration to the name, conflicting with how it works for non-MSVC.
* No more Win32 static build, use /MD everywhere.
* Building with /Zl to avoid defaultlib warnings.

In addition:
* I've reapplied "[sanitizer][asan][win] Intercept _strdup on Windows instead of strdup" which broke the previous static asan runtime. That runtime is gone now and this change is required for the strdup tests to work.
* I've modified the MSVC clang driver to support linking the correct asan libraries, including via defining _DLL (which triggers different defaultlibs and should result in the asan dll thunk being linked, along with the dll CRT (via defaultlib directives).
* I've made passing -static-libsan an error on windows, and made -shared-libsan the default. I'm not sure I did this correctly, or in the best way.
* Modified the test harnesses to add substitutions for the dynamic and static thunks and to make the library substitutions point to the dynamic asan runtime for all test configurations on windows. Both the static and dynamic windows test configurations remain, because they correspond to the static and dynamic CRT, not the static and dynamic asan runtime library.

### CODE REVIEW

**Review by vitalybuka — COMMENTED** [review]
This is long description, but it does not explain WHY.
Could you please add some explanation there?

**Comment by tru** [comment]
cc @sylvain-audi

**Comment by barcharcraz** [comment]
> This is long description, but it does not explain WHY. Could you please add some explanation there?

Sure, the reason is to greatly simplify how asan works on windows, and to fix a bunch of bugs with asan+static CRT. Before this change there was a bunch of code that would redirect calls from various dlls into the executable if the asan runtime was statically linked, this code has never worked all that well, and it was a nightmare to fix. Additionally, statically linked DLLs didn't work with the previous scheme and weak symbols didn't work with /MD.

**Comment by zmodem** [comment]
I think @rnk has the deepest knowledge here.

Just to double check, when you say "Eliminate the static asan runtime", you mean the use case it was supporting will still keep working, but in a different way right?

**Comment by barcharcraz** [comment]
> I think @rnk has the deepest knowledge here.
> 
> Just to double check, when you say "Eliminate the static asan runtime", you mean the use case it was supporting will still keep working, but in a different way right?

Yes. The static asan runtime didn't work very well in the first place, so this actually improves support for those use-cases. The static runtime is still supported on non-windows systems, however. 

The whole static linked asan runtime library is gone on windows, but you can still use the dynamic asan runtime with a statically linked CRT.

The core reasoning is that asan is a "only one allowed per process" type thing (you can't have one copy of the asan runtime handling a malloc while a different one handles the corresponding free). To get this on windows you really need to use a DLL or some really, really horrible hacks that essentially amounted to doing yourself what the loader already does for DLLs.

**Comment by mstorsjo** [comment]
> The core reasoning is that asan is a "only one allowed per process" type thing (you can't have one copy of the asan runtime handling a malloc while a different one handles the corresponding free).

Not to distract from the direction taken here (which I do agree seems reasonable, even if I haven't had time to look closer at the PR yet) - but, when using the static CRT in general, doesn't the same concern apply there as well? I.e. when using the static CRT, care must be taken that the same CRT instances does frees as did the allocation. But I guess there are other asan-related aspects that makes it even more of a mess.

Also secondly, when discussing these details - how the asan runtime is built in one, specific way, while it is used for applications that can use a different CRT configuration - would you like to chime in on https://github.com/microsoft/vcpkg/pull/34123#issuecomment-1760396486? There, some people involved questioned and weren't convinced that one part of a project (compiler-rt in llvm) should be allowed to override the CRT choice that is made for the whole vcpkg build. I tried to argue that compiler-rt is a very special case and doesn't interact with the rest of the libraries built in vcpkg like that, and compiler-rt is free to override the CRT to use internally.

**Comment by barcharcraz** [comment]
> > The core reasoning is that asan is a "only one allowed per process" type thing (you can't have one copy of the asan runtime handling a malloc while a different one handles the corresponding free).
> 
> Not to distract from the direction taken here (which I do agree seems reasonable, even if I haven't had time to look closer at the PR yet) - but, when using the static CRT in general, doesn't the same concern apply there as well? I.e. when using the static CRT, care must be taken that the same CRT instances does frees as did the allocation. But I guess there are other asan-related aspects that makes it even more of a mess.

Yes, it does, but my understanding is that this requirement is slightly weaker than what asan needs. As long as you _do_ match mallocs and frees things work fine, because the data that's shared is shared through kernel32.dll, ntdll.dll, or the kernel and page table. If asan allowed this it would need some way of mapping memory to the correct global structures, and some way of figuring out who gets to handle the shadow memory area and where to get the stack information for a given allocation upon an invalid access. We can't rely on "matching" allocations with invalid accesses because it's far from uncommon for a memory error to relate to memory allocated in another DLL. Maybe the static-linked asans could have some voting mechanism to settle on a "lead" copy of asan or something. It's not impossible but making things work well is a ton of work for something the loader can do anyway. Before this change we still relied on only one copy of asan being around, but the functions would be exported from the main exe with the "dll thunk" using GetProcAddress to call them. 

> Also secondly, when discussing these details - how the asan runtime is built in one, specific way, while it is used for applications that can use a different CRT configuration ....

It might be possible to support copies of the asan runtime built with the static CRT, it's just not that useful and we'd rather keep one configuration and spend time making it work perfectly for all instrumented programs.

**Comment by barcharcraz** [comment]
It occurs to me that this probably requires changes to the gyp build files.

**Comment by mstorsjo** [comment]
> > Not to distract from the direction taken here (which I do agree seems reasonable, even if I haven't had time to look closer at the PR yet) - but, when using the static CRT in general, doesn't the same concern apply there as well? I.e. when using the static CRT, care must be taken that the same CRT instances does frees as did the allocation. But I guess there are other asan-related aspects that makes it even more of a mess.
> 
> Yes, it does, but my understanding is that this requirement is slightly weaker than what asan needs. As long as you _do_ match mallocs and frees things work fine, because the data that's shared is shared through kernel32.dll, ntdll.dll, or the kernel and page table. If asan allowed this it would need some way of mapping memory to the correct global structures, and some way of figuring out who gets to handle the shadow memory area and where to get the stack information for a given allocation upon an invalid access. We can't rely on "matching" allocations with invalid accesses because it's far from uncommon for a memory error to relate to memory allocated in another DLL. Maybe the static-linked asans could have some voting mechanism to settle on a "lead" copy of asan or something. It's not impossible but making things work well is a ton of work for something the loader can do anyway. Before this change we still relied on only one copy of asan being around, but the functions would be exported from the main exe with the "dll thunk" using GetProcAddress to call them.

Ah, right, these are all very good explanations for why this is harder than the preexisting cases of multiple statically linked CRTs within the same process. Thanks!

> > Also secondly, when discussing these details - how the asan runtime is built in one, specific way, while it is used for applications that can use a different CRT configuration ....
> 
> It might be possible to support copies of the asan runtime built with the static CRT, it's just not that useful and we'd rather keep one configuration and spend time making it work perfectly for all instrumented programs.

Yeah, I totally agree, there's no practical need for allowing asan to be linked in another way. Especially as it works for both MSVC and mingw environments.

> It occurs to me that this probably requires changes to the gyp build files.

No, the extra build systems doesn't need to be updated in general, that's up to whoever maintains them to sync them afterwards. Regular contributors only need to make sure the cmake build works. (And unless one is able to test that the changes to the third party build systems actually is entirely right, it's usually better to leave them untouched, so it can be done cleanly in one commit, rather than do a half-broken update to the files.)

**Comment by barcharcraz** [comment]
The buildkite failure is just windows defender freaking out about one of cmake's test exes. Presumably it'll get sorted out next time it runs.

**Comment by rnk** [comment]
I'm short on time, so I can't respond to all the discussion here, but at a high level, I agree with this direction and it's something I discussed with Microsoft folks working on ASan in the past, so I'm happy to see it come together.

There is a great deal of build configuration complexity involved in supporting the static/dynamic variants. As mentioned, there should only ever be one copy of the ASan runtime active in a process address space. Making the static ASan build work properly required linking against different libraries depending on whether an EXE or a DLL was being produced, and this knowledge had to be encoded in every build system using ASan on Windows, since the linker is invoked directly, rather than running through the Clang driver. Ultimately, there is no real gain to be had from statically linking ASan. You ultimately have to make something equivalent to a shim import library for any DLL containing ASan instrumented code, so you might as well use the dynamic model and standardize.

I think the biggest risk here is, as usual, portability of weak symbol registration schemes. `__sanitizer_default_options` is registration-free, and there are some substantial advantages to making initialization order fiasco problems into duplicate symbol linker errors. In the end, the simplification benefits probably outweigh the portability issues.

I'll look at this some more and provide more feedback and hopefully other next steps.

**Review by rnk — COMMENTED** [review]
Is it feasible to create a migration path from the static runtime to the dynamic runtime? As written, this requires users to update their ASan builds at the same time that they take this update. Is it possible to leave the static runtime behind, create a migration path to the dynamic runtime (add all the new weak symbol registration functionality), push that, and then follow up by removing the static runtime 2-4 weeks later?

The sanitizer runtimes are not ABI-stable, so I think we're free to remove the static runtimes without leaving behind some kind of compatibility layer. This isn't libc++.

I also don't want to overindex on Google's or Chromium's needs. Many folks use the open source releases produced every six months, so to really accommodate those users, we'd have to delay the removal across a release boundary. I'm not advocating for that, but it's worth considering depending on the cost.

I believe the first four patches can land separately if we agree on the direction.

**Comment by barcharcraz** [comment]
> Is it feasible to create a migration path from the static runtime to the dynamic runtime? As written, this requires users to update their ASan builds at the same time that they take this update. Is it possible to leave the static runtime behind, create a migration path to the dynamic runtime (add all the new weak symbol registration functionality), push that, and then follow up by removing the static runtime 2-4 weeks later?

The new symbol registration mechanism depends on how the loader imports multiple DLLexports of the same symbol, maybe it could be adapted to work with the static runtime, but it's a bunch of work for something that will be immediately removed.

We could also just ship a copy of the dynamic runtime import library under the same name as the old static runtime, perhaps emitting a warning someplace to let people know what's going on. 

So, for context: MSVC's link.exe uses a special /INFERASANLIBS option that picks the correct asan library automatically, and this PR should (hopefully, I haven't tested it yet) make that flag work with clang's version of asan as well, at least for release builds (MSVC ships with debug builds of asan as well, they're still linked to the release-dll variant of the CRT, they just have optimization turned off). When you pass /fsanitize=address to cl.exe it adds the equivalent of `#pragma comment(linker, "/INFERASANLIBS")` to the object files. Link.exe sees this and picks the right libraries by seeing which crt it was going to link and inserting the asan libraries before the crt in the link order. 

I'm not quite sure I understand the problem scenario. Is the trouble that if you're building with mingw's gcc then you need to invoke the linker manually because gcc doesn't know how to deal with link.exe (and you want to use link.exe over ld.exe because ld is slow and presumably not that well maintained on windows?). I think clang can invoke link.exe for you, just like any other linker, or is that only lld-link? cl can also invoke link for you, although I understand going through the compiler like that is less common on windows that on other platforms.

**Comment by vitalybuka** [comment]
> @vitalybuka Here's a more detailed explination on the motivations behind this change from @amyw-msft, who is the original author of these changes on our side. (And who I'll add as a co-author using fixup commits)
> 
> https://devblogs.microsoft.com/cppblog/msvc-address-sanitizer-one-dll-for-all-runtime-configurations/

Thanks. Would be possible to extract summary paragraph and put on top of the main  patch, with this URL.

Also, to submit, it would be nice to do with smaller patches with some time between them to let bots test them separately. Or it's going to be annoying to revert/reland such large patch.

Almost every item from description looks like a candidate for a patch.

Relaxing `// CHECK: #{{[1-3]}}` could be done in a separate patch.

**Comment by barcharcraz** [comment]
> > @vitalybuka Here's a more detailed explination on the motivations behind this change from @amyw-msft, who is the original author of these changes on our side. (And who I'll add as a co-author using fixup commits)
> > https://devblogs.microsoft.com/cppblog/msvc-address-sanitizer-one-dll-for-all-runtime-configurations/
> 
> Thanks. Would be possible to extract summary paragraph and put on top of the main patch, with this URL.
> 
> Also, to submit, it would be nice to do with smaller patches with some time between them to let bots test them separately. Or it's going to be annoying to revert/reland such large patch.
> 
> Almost every item from description looks like a candidate for a patch.
> 
> Relaxing `// CHECK: #{{[1-3]}}` could be done in a separate patch.

OK I'll extract at least the first few commits into their own PR, along with the test relaxations. The items in the description are harder to extract without breaking all the tests.

**Comment by barcharcraz** [comment]
Test changes that don't require anything else in https://github.com/llvm/llvm-project/pull/85274.

**Comment by barcharcraz** [comment]
@rnk @vitalybuka

This should be ready for a more in-depth review (I'll stop force-pushing)

The remaining commits here all depend on each-other in some way. Splitting off any more commits would mean introducing bugs into main until other parts of this are merged.

File: compiler-rt/test/asan/TestCases/Darwin/interface_symbols_darwin.cpp [hunk 1]
```diff
@@ -35,6 +35,9 @@
 // RUN:  %p/../../../../lib/sanitizer_common/sanitizer_coverage_interface.inc     \
 // RUN:  | grep -e "INTERFACE_\(WEAK_\)\?FUNCTION"                                \
 // RUN:  | grep -v "__sanitizer_weak_hook"                                        \
+// RUN:  | grep -v "__sanitizer_override_function"                                \
```
Comment (vitalybuka) [inline compiler-rt/test/asan/TestCases/Darwin/interface_symbols_darwin.cpp:38]: can we move related changes into a separate patch

**Comment by vitalybuka** [comment]
It's LGTM, but it's going to be annoying to merge.
I bet this patch has several reasons to be reverted, and will be reverted multiple times.
It would be nice to try to split into smaller parts.

**Comment by MaskRay** [comment]
> It's LGTM, but it's going to be annoying to merge. I bet this patch has several reasons to be reverted, and will be reverted multiple times. It would be nice to try to split into smaller parts.

I agree.

It'd be nice if Chromium folks can test this before this is merged.

File: compiler-rt/test/asan/TestCases/Darwin/interface_symbols_darwin.cpp (hunk 1 above)
Comment (barcharcraz) [reply compiler-rt/test/asan/TestCases/Darwin/interface_symbols_darwin.cpp:38]: It would cause test failures on linux/darwin. These tests grep the interface files for "expected" imports and do not run the preprocessor on them, so the new interface symbols on windows need to be excluded.
